### PR TITLE
Maximum Retry Limit for Firmware Upgrades

### PR DESCRIPTION
**Pull request by dmccoystephenson**
# PR Details
### Problem
The current firmware manager continuously attempts to upgrade the firmware for an RSU, regardless of how many failures occur. This behavior can clog the upgrade queue and waste CPU resources, especially when an RSU is unreachable.

### Solution
To address this, a new environment variable has been introduced to set a maximum retry limit for firmware upgrades per RSU. Each time a firmware upgrade fails, the system will increase a 'consecutive failure count' for that RSU. Once the count reaches the maximum retry limit, the target_firmware_version in the rsus table will be updated to match the current firmware_version, effectively halting further upgrades for that RSU. The failure will also be logged in a new Postgres table.

Additionally, the firmware manager will now skip upgrade attempts for RSUs that have returned a negative ping result.

## How Has This Been Tested?
Unit tests have been implemented to verify function calls and behavior of methods. Additionally, the firmware manager was spun up in a local docker network with an ubuntu container serving as a mock RSU with a `signedUpgrade.sh` script in the /bin directory that just printed a hard-coded failure alert. The IP address of the ubuntu container was set as the value for the IP address of a sample RSU and three firmware upgrades were executed manually by calling the relevant endpoint.

The firmware manager modified the `target_firmware_version` to `firmware_version` as expected and logged the incident to the 'max_retry_limit_reached_instances' table.

image

## Types of changes
- [ ] Defect fix (non-breaking change that fixes an issue)
- [ x ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:
- [ x ] My changes require new environment variables:
  - [ / ] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [ ] My changes require updates to the documentation:
  - [ ] I have updated the documentation accordingly.
- [ x ] My changes require updates and/or additions to the unit tests:
  - [ x ] I have modified/added tests to cover my changes.
- [ x ] All existing tests pass.

### CODE REVIEW

**Review by drewjj — APPROVED** [review]
This looks great! A lot for such a simple check (conceptually) but I think the corner cases covered are all necessary additions so I have no complaints. All unit tests pass.

**Review by mwodahl — APPROVED** [review]
This all looks good to me!

File: services/addons/images/firmware_manager/firmware_manager.py [hunk 1]
```diff
@@ -271,6 +302,53 @@ def check_for_upgrades():
...
+    return latest_ping_successful
+
+
+def increment_consecutive_failure_count_for_rsu(rsu_ip):
+    if not does_consecutive_failure_count_exist_for_rsu(rsu_ip):
+        insertion_query = f"insert into consecutive_firmware_upgrade_failures (rsu_id, consecutive_failures) values ((select rsu_id from rsus where ipv4_address='{rsu_ip}'), 1)"
+        pgquery.write_db(insertion_query)
+    else:
+        increment_query = f"update consecutive_firmware_upgrade_failures set consecutive_failures=consecutive_failures+1 where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}')"
+        pgquery.write_db(increment_query)
```
Comment (mcook42) [inline services/addons/images/firmware_manager/firmware_manager.py]: question (non-blocking): I'm new-ish to this codebase and a little rusty with my Postgres, but I'm wondering if there's a reason we don't use an [Upsert ](https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/)? If we used an upsert statement we could combine the logic in `does_consecutive_failure_count_exist_for_rsu`, the if-block, and the else-block in this function into one call to the database.

File: resources/sql_scripts/update_scripts/firmware_max_retry_limit_update.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+CREATE TABLE IF NOT EXISTS public.max_retry_limit_reached_instances
+(
+   rsu_id integer NOT NULL,
+   reached_at timestamp without time zone NOT NULL,
+   CONSTRAINT max_retry_limit_reached_instances_pkey PRIMARY KEY (rsu_id, reached_at),
+   CONSTRAINT fk_rsu_id FOREIGN KEY (rsu_id)
+      REFERENCES public.rsus (rsu_id) MATCH SIMPLE
+      ON UPDATE NO ACTION
+      ON DELETE NO ACTION
+);
```
Comment (mcook42) [inline resources/sql_scripts/update_scripts/firmware_max_retry_limit_update.sql]: question/suggestion (non-blocking): could these two tables be combined into one like the following? I'm unsure about the context and the database design standards, but combining the two would reduce data duplication and (possibly) make the debugging developer's life a little easier since they wouldn't need to query multiple tables for the information.

```suggestion
CREATE TABLE IF NOT EXISTS public.consecutive_firmware_upgrade_failures
(
   rsu_id integer NOT NULL,
   consecutive_failures integer NOT NULL,
   last_update_attempted_at timestamp without time zone NOT NULL,
   CONSTRAINT consecutive_firmware_upgrade_failures_pkey PRIMARY KEY (rsu_id),
   CONSTRAINT fk_rsu_id FOREIGN KEY (rsu_id)
      REFERENCES public.rsus (rsu_id) MATCH SIMPLE
      ON UPDATE NO ACTION
      ON DELETE NO ACTION
);
```

The table could be used in the code in a similar way as it is currently. The queries would just need to check `consecutive_failures` against the configured `FW_UPGRADE_MAX_RETRY_LIMIT`. Obviously, other code changes would be needed, but this is the most obvious.

File: services/addons/tests/firmware_manager/test_firmware_manager.py [hunk 3]
```diff
@@ -595,17 +779,180 @@ def test_check_for_upgrades(mock_upgrade_limit, mock_stfq, mock_logging):
...
+    "addons.images.firmware_manager.firmware_manager.does_consecutive_failure_count_exist_for_rsu"
+)
+@patch("addons.images.firmware_manager.firmware_manager.pgquery.write_db")
+def test_increment_consecutive_failure_count_for_rsu_record_exists(
+    mock_write_db, mock_does_consecutive_failure_count_exist_for_rsu
+):
+    # prepare
+    mock_does_consecutive_failure_count_exist_for_rsu.return_value = True
+    rsu_ip = "8.8.8.8"
+    expected_query = "update consecutive_firmware_upgrade_failures set consecutive_failures=consecutive_failures+1 where rsu_id=(select rsu_id from rsus where ipv4_address='8.8.8.8')"
```
Comment (mcook42) [inline services/addons/tests/firmware_manager/test_firmware_manager.py]: question(non-blocking): This a general question about the tests in this file. I'm unfamiliar with testing for which query was executed. I'm more used to testing whether the correct data was returned by a function. What does testing for the specific query we run get us?

File: resources/sql_scripts/update_scripts/firmware_max_retry_limit_update.sql (hunk 2 above)
Comment (dmccoystephenson) [reply resources/sql_scripts/update_scripts/firmware_max_retry_limit_update.sql]: One table tracks consecutive failure counts to help decide whether to continue upgrade attempts. The second table, `public.max_retry_limit_reached_instances`, serves as a log for instances where the maximum retry limit has been reached. Developers can reference this log for additional context. I don't think combining these two tables would preserve their distinct purposes.

File: services/addons/tests/firmware_manager/test_firmware_manager.py (hunk 3 above)
Comment (dmccoystephenson) [reply services/addons/tests/firmware_manager/test_firmware_manager.py]: Unit tests should depend solely on the class being tested. If we wanted to test with real data, we would need a live database running in the background. This test checks that the correct queries are executed with the expected parameters when the method is called. By validating this behavior without a live database, the test ensures the class logic works correctly in isolation.

File: services/addons/images/firmware_manager/firmware_manager.py (hunk 1 above)
Comment (dmccoystephenson) [reply services/addons/images/firmware_manager/firmware_manager.py]: I really like the idea of using ON CONFLICT to reduce the number of database queries here. I've opened a PR to implement your suggestion here: https://github.com/CDOT-CV/jpo-cvmanager/pull/112

Comment (dmccoystephenson) [reply services/addons/images/firmware_manager/firmware_manager.py]: Addressed!

File: services/addons/images/firmware_manager/firmware_manager.py [hunk 4]
```diff
@@ -213,6 +225,18 @@ def firmware_upgrade_completed():
                     ),
                     500,
                 )
+        else:
+            increment_consecutive_failure_count_for_rsu(request_args["rsu_ip"])
+            if is_rsu_at_max_retries_limit(request_args["rsu_ip"]):
+                logging.error(f"RSU {request_args['rsu_ip']} has reached the maximum number of upgrade retries. Setting target_firmware_version to firmware_version and resetting consecutive failures count.")
```
Comment (mcook42) [inline services/addons/images/firmware_manager/firmware_manager.py]: question(non-blocking): does this log mean we're updating the target_firmware_version to the current version (like the version already installed on the rsu) or to the firmware_version we tried but failed to install (the newest firmware version)? I'm pretty sure it's meant to be the former, but for some reason my brain is having a hard time feeling certain about what this log means 😅

File: services/addons/images/firmware_manager/firmware_manager.py [hunk 5]
```diff
@@ -271,6 +302,38 @@ def check_for_upgrades():
...
+    upsert_query = f"insert into consecutive_firmware_upgrade_failures (rsu_id, consecutive_failures) values ((select rsu_id from rsus where ipv4_address='{rsu_ip}'), 0) on conflict (rsu_id) do update set consecutive_failures=0"
+    pgquery.write_db(upsert_query)
+
+
+def is_rsu_at_max_retries_limit(rsu_ip):
+    # get max retries from environment variable
+    max_retries = int(os.environ.get("FW_UPGRADE_MAX_RETRY_LIMIT", "3"))
+    consecutive_failures = pgquery.query_db(
+        f"select consecutive_failures from consecutive_firmware_upgrade_failures where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}')"
+    )[0][0]
```
Comment (mcook42) [inline services/addons/images/firmware_manager/firmware_manager.py]: question(blocking): I may be overly cautious with this, but what happens if we get an empty array back from `query_db`? Will we get an "index out of bounds exception" when we try to access `[0][0]` on the array, or is there a different expected behavior?

File: services/addons/images/firmware_manager/firmware_manager.py (hunk 4 above)
Comment (dmccoystephenson) [reply services/addons/images/firmware_manager/firmware_manager.py]: The `rsus` table contains two columns related to firmware: `firmware_version` and `target_firmware_version`. The `firmware_version` column shows the current firmware version installed on the RSU, while the `target_firmware_version` indicates the desired firmware version we want to install.

By setting `target_firmware_version` to match the current value of the `firmware_version` column, we are instructing the firmware manager to keep the RSU on its existing firmware and prevent any upgrades.

File: services/addons/images/firmware_manager/firmware_manager.py (hunk 5 above)
Comment (dmccoystephenson) [reply services/addons/images/firmware_manager/firmware_manager.py]: Nice catch! An index error could happen in this case. I've added a try/catch block, so if an index error occurs while reading the database response, the method will now return false.

File: services/addons/images/firmware_manager/firmware_manager.py [hunk 6]
```diff
@@ -259,18 +299,67 @@ def check_for_upgrades():
...
+                f"Firmware upgrade successfully started for '{rsu["ipv4_address"]}'"
+            )
 
         # Start any processes that can be started
         start_tasks_from_queue()
 
 
+def was_latest_ping_successful_for_rsu(rsu_ip):
+    query = f"select result from ping where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}') order by timestamp desc limit 1"
+    latest_ping_successful = pgquery.query_db(query)[0][0]
```
Comment (mcook42) [inline services/addons/images/firmware_manager/firmware_manager.py]: issue: if we receive an empty array from `query_db()` we're at risk of getting and index out of bound exception.  We should first check the length of the results, then try to access the data within them. So something like the following would work
```suggestion
    pings = pgquery.query_db(query)[0][0]
    if len(pings) == 0 or len(pings[0]) == 0:
        return False
    latest_ping_successful = pings[0][0]
```

Comment (dmccoystephenson) [reply services/addons/images/firmware_manager/firmware_manager.py]: Addressed!

**Review by mcook42 — APPROVED** [review]

File: services/addons/images/firmware_manager/firmware_manager.py [hunk 7]
```diff
@@ -259,18 +299,66 @@ def check_for_upgrades():
...
+        f"select consecutive_failures from consecutive_firmware_upgrade_failures where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}')"
+    )
+    if len(query_result) == 0 or len(query_result[0]) == 0:
+        # no failures have been recorded for this RSU, so it cannot be at the limit
+        return False
+    consecutive_failures = query_result[0][0]
+    return consecutive_failures >= max_retries
+
+
+def log_max_retries_reached_incident_for_rsu_to_postgres(rsu_ip):
```
Comment (payneBrandon) [inline services/addons/images/firmware_manager/firmware_manager.py]: Do you think it'd make sense to also log the attempted firmare version that was causing the issue? It's possible we'd see repeated failures from various versions and might be nice to have that version in here to know exactly which one failed

**Review by payneBrandon — CHANGES_REQUESTED** [review]
Just the one question, looking good!

File: services/addons/images/firmware_manager/firmware_manager.py (hunk 7 above)
Comment (dmccoystephenson) [reply services/addons/images/firmware_manager/firmware_manager.py]: Yeah I think this makes sense. I've modified the max retry limit reached instance tracking to include the firmware id corresponding to the target firmware version at the time of upgrading.

**Review by payneBrandon — APPROVED** [review]
lgtm!
